### ISSUE TITLE
Unable to serialize generics in Scala 

### ISSUE THREAD

**Issue by inmyth** [ngs-doo/dsl-json#100]
I have been unable to serialize a generic class using [GenericTest](https://github.com/ngs-doo/dsl-json/blob/master/tests-java8/src/test/java/com/dslplatform/json/GenericTest.java) as reference. I have below a very simple class in Scala. To imitate the Java test code I don't use Option or case class. 

```
  @CompiledJson
  class Abc[A] {
    var test: A = _
  }

  val dslJson = new DslJson[Any](new DslJson.Settings()
                .allowArrayFormat(true).includeServiceLoader())
  val writer = dslJson.newWriter
  var model = new Abc[String]
  model.test = "aaa"
  val t = new TypeDefinition[Abc[String]]() {}.`type`
  val os = new ByteArrayOutputStream
  writer.reset(os)
  val res = dslJson.serialize(writer, t, model)     
  writer.flush()
```  

`res` always returns false signifying failed serialization. Perhaps you could help point out the problem ? Thank you.

**Comment by zapov** [comment]
Unfortunately, thats currently expected behavior.
Generics are supported only via annotation processor.
Scala version currently uses only reflection features.
Scala tooling is not really annotation processor friendly, and while technically you could run it, in practice I have not seen people do it.

Ideally Scala version will get some kind of compilation support... but community should help out with that.

**Comment by inmyth** [comment]
I see. I guess at the time being, generics are only supported in Java version ?

**Comment by zapov** [comment]
Yes

**Comment by inmyth** [comment]
So I tried recoding it in Java but it still didn't work. Finally I copied and ran [GenericTest](https://github.com/ngs-doo/dsl-json/blob/master/tests-java8/src/test/java/com/dslplatform/json/GenericTest.java) but all tests failed.  I was using dsl-json-java8, version: 1.8.3 (version 1.8.4 was not yet available at Maven). This is one result I get from `testSerializeAndDeserializeGeneric` . Basically `dslJson.serialize` would still return false and the resulting json returns an empty string. 

```
java.io.IOException: Unexpected end of JSON input

	at com.dslplatform.json.JsonReader.read(JsonReader.java:304)
	at com.dslplatform.json.JsonReader.getNextToken(JsonReader.java:747)
	at com.dslplatform.json.DslJson.deserialize(DslJson.java:1550)
	at com.mypackage.json.dsl.GenericTest.testSerializeAndDeserializeGeneric(GenericTest.java:67)
```

**Comment by zapov** [comment]
If you write your example as a test in GenericTest does it work then? If not, can you make a PR?

I suspect that you need to initialize library via new `DslJson<>(Settings.withRuntime())` before it behaves as you expect.

**Comment by inmyth** [comment]
I copied the file without changing anything. DslJson is initialialized like this
      
    private final DslJson<Object> dslJson = new DslJson<>(new DslJson.Settings<>()
    .allowArrayFormat(true).includeServiceLoader());

And I tried cloning the project. It failed to build with the same reason as when I did it with my own test. Basically `dsl-json` version 1.8.4 is not available and artifact `dsl-json-integration-tests-java8-external-models` doesn't exist. As the last version was pushed very recent like a week ago, maybe it has something to do with it ?

**Comment by zapov** [comment]
dsl-json 1.8.4 is not released, but you should be able to build it locally by going to root project and running 

    mvn clean install

That should build the entire project including the external-models.
What happens if you do that?

**Comment by inmyth** [comment]
Still failed with error
```
DSL Platform JSON (core) 1.8.4 ..................... FAILURE [ 10.291 s]
java.lang.NoClassDefFoundError: org/junit/runner/notification/RunNotifier
	at java.lang.Class.getDeclaredConstructors0(Native Method)
Caused by: java.lang.ClassNotFoundException: org.junit.runner.notification.RunNotifier
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
```
though test-compile works.

**Comment by wadejensen** [comment]
Would be very keen for generics support.

**Comment by zapov** [comment]
Yes, it would be nice to resolve such minor TODOs: https://github.com/ngs-doo/dsl-json/blob/master/scala/src/main/scala/com/dslplatform/json/runtime/ScalaClassAnalyzer.scala#L20

I don't have much time ATM, but I suspect it would not be complicated to add that support for reflection.

**Comment by zapov** [comment]
There is now basic support for generics in Scala: https://github.com/ngs-doo/dsl-json/blob/master/scala/src/test/scala/com/dslplatform/json/GenericsTest.scala

Will be included in 1.8.5

**Comment by zapov** [comment]
v1.8.5 released
